src/cdi_ndi/NDI_TN.cc

```diff
@@ -0,0 +1,279 @@
+
+      //! Set NDI to reaction product
+      SAFE_NDI(
+          NDI2_set_option(dataset_handle, NDI_CURR_PART, product_zaid.c_str()));
+
+      //! Get number of temperature support points (this can depend on reaction
+      //! product)
+      int num_product_temps =
+          NDI2_get_size(dataset_handle, NDI_EDIST_TEMPS, &ndi_size_error);
+      Require(ndi_size_error == 0);
+      product_temperatures[n].resize(num_product_temps);
+      product_distributions[n].resize(num_product_temps);
+
+      //! Get temperature support points
+      SAFE_NDI(NDI2_get_float64_vec(
+          dataset_handle, NDI_TEMPS, product_temperatures[n].data(),
+          static_cast<int>(product_temperatures[n].size())));
+      // MeV -> keV
+      for (auto &temperature : product_temperatures[n]) {
+        temperature *= 1000.;
+      }
+
+      //! Get the number of interp regions... for now just throw an exception if
+      //! this is not equal to 1
+      int num_edist_interp_regions =
+          NDI2_get_size(dataset_handle, NDI_EDIST_INTERP_REG, &ndi_size_error);
+      Require(ndi_size_error == 0);
+      Insist(num_edist_interp_regions == 1,
+             "Only 1 edist interp region supported!");
+
+      // Loop over temperatures
+      for (size_t m = 0; m < product_temperatures[n].size(); m++) {
+        std::ostringstream temp_stream;
+        temp_stream << product_temperatures[n][m] / 1000; // keV -> MeV
+        SAFE_NDI(NDI2_set_option(dataset_handle, NDI_TEMP,
+                                 temp_stream.str().c_str()));
+
+        product_distributions[n][m].resize(num_groups);
+        SAFE_NDI(NDI2_get_float64_vec(
+            dataset_handle, NDI_EDIST, product_distributions[n][m].data(),
+            static_cast<int>(product_distributions[n][m].size())));
+      }
+    }
+  }
+
+  //! Close datafile
+  SAFE_NDI(NDI2_close_gendir(gendir_handle));
+}
+
+//----------------------------------------------------------------------------//
+/*!
+ * \brief Sample reaction product energy distribution at fixed temperature.
+ * \param[in] product_zaid ZAID of reaction product to sample
+ * \param[in] temperature of plasma (keV)
+ * \return Sampled reaction product energy (keV)
+ */
+double NDI_TN::sample_distribution(const int product_zaid,
+                                   const double temperature) {
+  const int product_index = product_zaid_to_index[product_zaid];
+
+  Require(temperature > product_temperatures[product_index].front());
+  Require(temperature < product_temperatures[product_index].back());
+
+  auto temp_1 =
+      std::upper_bound(product_temperatures[product_index].begin(),
+                       product_temperatures[product_index].end(), temperature);
+  uint32_t index_1 = static_cast<uint32_t>(
+      temp_1 - product_temperatures[product_index].begin());
+  uint32_t index_0 = index_1 - 1;
+  double temp_0 = product_temperatures[product_index][index_0];
+  double fac = 1. - (temperature - temp_0) / (*temp_1 - temp_0);
```

Comment:
* Check for division by zero? or division by small?
* Check 0 < fac < 1?

Reply:
Both done (except 0 =< fac <= 1 since we are allowed to access data directly on a support point)